crates/ruma-events/src/call/member/member_state_key.rs

```diff
@@ -0,0 +1,223 @@
+            CallMemberStateKeyEnum::User(u) => u,
+        }
+    }
+
+    /// Returns the device id in this state key (if available)
+    /// (This is a cheap operations. The id is already type checked on initialization. And does
+    /// only returns a reference to an existing OwnedDeviceId.)
+    pub fn device_id(&self) -> Option<&DeviceId> {
+        match &self.key {
+            CallMemberStateKeyEnum::UnderscoreUserDevice(_, d) => Some(d),
+            CallMemberStateKeyEnum::UserDevice(_, d) => Some(d),
+            CallMemberStateKeyEnum::User(_) => None,
+        }
+    }
+}
+
+impl AsRef<str> for CallMemberStateKey {
+    fn as_ref(&self) -> &str {
+        &self.raw
+    }
+}
+
+impl From<CallMemberStateKeyEnum> for CallMemberStateKey {
+    fn from(value: CallMemberStateKeyEnum) -> Self {
+        let raw = value.to_string().into();
+        Self { key: value, raw }
+    }
+}
+
+impl FromStr for CallMemberStateKey {
+    type Err = KeyParseError;
+
+    fn from_str(state_key: &str) -> Result<Self, Self::Err> {
+        // Intentionally do not use CallMemberStateKeyEnum.into since this would reconstruct the
+        // state key string.
+        Ok(Self { key: CallMemberStateKeyEnum::from_str(state_key)?, raw: state_key.into() })
+    }
+}
+
+impl<'de> Deserialize<'de> for CallMemberStateKey {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = ruma_common::serde::deserialize_cow_str(deserializer)?;
+        Self::from_str(&s).map_err(|err| de::Error::invalid_value(Unexpected::Str(&s), &err))
+    }
+}
+
+impl Serialize for CallMemberStateKey {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(self.as_ref())
+    }
+}
+
+/// This enum represents all possible formats for a call member event state key.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+enum CallMemberStateKeyEnum {
+    UnderscoreUserDevice(OwnedUserId, OwnedDeviceId),
+    UserDevice(OwnedUserId, OwnedDeviceId),
+    User(OwnedUserId),
+}
+
+impl CallMemberStateKeyEnum {
+    fn new(user_id: OwnedUserId, device_id: Option<OwnedDeviceId>, underscore: bool) -> Self {
+        match (device_id, underscore) {
+            (Some(device_id), true) => {
+                CallMemberStateKeyEnum::UnderscoreUserDevice(user_id, device_id)
+            }
+            (Some(device_id), false) => CallMemberStateKeyEnum::UserDevice(user_id, device_id),
+            (None, _) => CallMemberStateKeyEnum::User(user_id),
+        }
+    }
+}
+
+impl std::fmt::Display for CallMemberStateKeyEnum {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match &self {
+            CallMemberStateKeyEnum::UnderscoreUserDevice(u, d) => write!(f, "_{}_{}", u, d),
+            CallMemberStateKeyEnum::UserDevice(u, d) => write!(f, "{}_{}", u, d).into(),
```

Comment:
Prefer named parameters:

```suggestion
            CallMemberStateKeyEnum::UnderscoreUserDevice(u, d) => write!(f, "_{u}_{d}"),
            CallMemberStateKeyEnum::UserDevice(u, d) => write!(f, "{u}_{d}").into(),
```